### PR TITLE
compiler: Fix attribute builder

### PR DESCRIPTION
**Pull request by math-fehr**
Fix `Attribute.build`, which failed if too many arguments were given.
It also failed when using `|` rather than `Union`.

Also update the builder tests, and fix obviously wrong ones.

### CODE REVIEW

File: xdsl/utils/hints.py [hunk 1]
```diff
@@ -35,7 +36,7 @@ def is_satisfying_hint(arg: Any, hint: type[_T]) -> TypeGuard[_T]:
             and is_satisfying_hint(value, value_hint)
             for key, value in arg_dict.items())
 
-    if get_origin(hint) is Union:
+    if get_origin(hint) is Union or get_origin(hint) is UnionType:
```
Comment (georgebisbas) [inline xdsl/utils/hints.py:39]: if get_origin(hint) is in/any {.....} ?

**Review by georgebisbas — APPROVED** [review]

File: xdsl/utils/exceptions.py [hunk 2]
```diff
@@ -1,6 +1,30 @@
...
 
 class VerifyException(DiagnosticException):
-    ...
+    pass
+
+
+@dataclass
+class BuilderNotFoundException(Exception):
+    """
+    Exception raised when no builders are not found for a given attribute type
```
Comment (webmiche) [inline xdsl/utils/exceptions.py:22]:
```suggestion
    Exception raised when no builders are found for a given attribute type
```

**Review by webmiche — APPROVED** [review]
LGTM, modulo minor comments.

File: tests/test_attribute_builder.py [hunk 3]
```diff
@@ -52,10 +53,43 @@ def test_one_builder_builder():
...
+    @staticmethod
+    def parse_parameter(parser: Parser) -> str:
+        raise NotImplementedError()
+
+    @staticmethod
+    def print_parameter(data: str, printer: Printer) -> None:
+        raise NotImplementedError()
+
+
+def test_one_builder_2_not_enough_args():
```
Comment (georgebisbas) [inline tests/test_attribute_builder.py:83]: Would avoid using "2"

File: tests/test_attribute_builder.py [hunk 4]
```diff
@@ -95,68 +129,83 @@ def test_two_builders_second_builder():
...
 
-def builder_default_arg_arg():
-    attr = BuilderDefaultArgAttr.build(4, 2)
-    assert attr == BuilderDefaultArgAttr("42")
+def test_builder_default_args_all_args():
+    attr = BuilderDefaultArgAttr.build(4, 5, 6)
+    assert attr == BuilderDefaultArgAttr("4, 5, 6")
+
+
+def test_builder_default_args_too_many_args():
```
Comment (georgebisbas) [inline tests/test_attribute_builder.py:176]: try to drop "too" from the name

Comment (math-fehr) [reply tests/test_attribute_builder.py]: Done.
I'm actually curious, why is this bad practice?

**Review by webmiche — APPROVED** [review]

**Comment by webmiche** [comment]
Can we merge? @math-fehr

**Pull request merged**